### ISSUE TITLE
Document.Close

### ISSUE THREAD

**Issue by Tyree** [erfajo/OrchidForDynamo#46]
Looking at older usage scenarios for Orchid, it appears that nodes like setting a parameter on a family type (Parameter.SetValue) used to return the family document. Now it returns the Parameter.

I am opening the document, creating several types, and setting parameters on each and THEN I want to save and close the document. How do I get that to happen in sequence? How do I get the document object back again AFTER the parameters have been set?

Thanks!

**Comment by erfajo** [comment]
use a WaitFor node... when the output from Parameter.SetValue is complete, then make this the input to wait for, and set the documents as the pass data.
I made the change, since the parameters could be used for other purposes, and if you wanted to pass documents, then was it easy to add the sequence described.

**Comment by Tyree** [comment]
Great. Thanks! I found the Datastream.Await node. How does this work when setting multiple params on multiple new types? Will the Await trigger after the first param is set or not until all params are set?

**Comment by erfajo** [comment]
could you drop the graph file, then I could take a look at it... rename the .dyn file to .dyn.txt and then upload it

**Comment by Tyree** [comment]
Sure thing. Thanks!
[AP_Note Blocks_ZMASTER_Import from Excel.dyn.txt](https://github.com/erfajo/OrchidForDynamo/files/3751782/AP_Note.Blocks_ZMASTER_Import.from.Excel.dyn.txt)

**Comment by erfajo** [comment]
I have made some few changes and some optimizing you can use...
[AP_Note.Blocks_ZMASTER_Import.from.Excel.dyn.txt](https://github.com/erfajo/OrchidForDynamo/files/3752082/AP_Note.Blocks_ZMASTER_Import.from.Excel.dyn.txt)

**Comment by Tyree** [comment]
I'm still very new to Dynamo and visual/graph programming. Thanks for taking the time to help me out. I'm going to go over your changes step by step. First, I don't get the same result from the Remove Null Values group. It needs to remove the indexes from all three lists if one list has a null value at that index. Not just all null values from all lists. Does that make sense?
![removenullvalues](https://user-images.githubusercontent.com/5395363/67231246-f0e78380-f40c-11e9-86a4-28081649a5f8.PNG)

**Comment by Tyree** [comment]
I like how you have streamlined the "Create Type Titles" group. I am getting a warning each time on the substring Code Block, though. "Dereferencing a non-pointer." Any idea what that means? If I sub in my old substring Code Block I don't get that error. But the code appears to do the exact same thing. You just use the ternary version of the if/else.
![substring_error](https://user-images.githubusercontent.com/5395363/67232975-5ee17a00-f410-11e9-84b1-2dc717f9d63c.PNG)

**Comment by Tyree** [comment]
The real trick, and where I am stuck is on the end. The Set Parameters on New Types group. The script creates the new types, but it only uses the first set of values from the 3 value lists on ALL the new types. There are 327 types and 327 of each value, but I can't figure out how to get them to sync up and assign the values to the correct type. I know that's well outside of what you're trying to help me with on Orchid, but I know you know way more than I do about Dynamo. :-)
![all](https://user-images.githubusercontent.com/5395363/67233286-eaf3a180-f410-11e9-98d7-bdc0ada8cb89.png)

**Comment by huvec** [comment]
I am not sure if i should set up a new issue or continue with this as I am having hte same issue. I can setup as a new issue if required.
I have seeing the same issue for the set parameters node. I am updating the type parameters into RFA files. The families have different number of types in each.  The output is picking up the first value of a list.

Also are you able to explain how the parameter.setvalue will work with different number of files and types. In the case of the image below, the 1st family has 1 type while the second family has 3 types.
![image](https://user-images.githubusercontent.com/16100298/67250369-96f7c400-f4c7-11e9-8d5c-748a9dae2eb4.png)

**Comment by Tyree** [comment]
I tried a different way of grouping and feeding the data to the set params node and I have yet to be able to get the script to run. I've had to kill the Revit process a few times. Just says "Run started..." and never finishes. :-/
[AP_Note.Blocks_ZMASTER_Import.from.Excel.dyn.txt](https://github.com/erfajo/OrchidForDynamo/files/3755365/AP_Note.Blocks_ZMASTER_Import.from.Excel.dyn.txt)

**Comment by erfajo** [comment]
sorry I have been busy. I can look into it later... but the error "Dereferencing a non-pointer." is most likely the null value error, as you pointet in an earliere topic, that the entire sublist should be removed... then stick to your way of doing it.... I will return later :-)

**Comment by Tyree** [comment]
No worries, @erfajo . I worked around this temporarily by doing the work inside a project file and then exporting the family. Definitely not the way I'd like to do it, but it works for now.
I also have a thread on DynamoBIM for this:
https://forum.dynamobim.com/t/creating-multiple-family-types-then-assigning-param-values-to-them/42130

**Comment by Tyree** [comment]
Oh, one thing I did notice. I shrank my sample Excel file down to just 30 or so entries to see what would happen if I let it run for a while. It did finish, but it puts in "System.Collections.ArrayList" for the param values. So, there's still something about how I'm packaging up the param values that Parameter.SetValue isn't liking.

**Comment by Tyree** [comment]
So, I don't want to say this too loud, but I think I figured it out. I've tested it a few times and it seems reliable.

I just had the change the lacing on the Parameter.SetValue node to longest from auto. Now it matches the indexes up properly.

**Comment by erfajo** [comment]
sounds good... and yes lacing is also a good thing to be familiar with :-)

sounds like this issue can be closed :-)
